[AO][bugfix] Fixing FX QAT but for untraceable modules

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74277

Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by skipping the children of untraceable modules during
propagate_qconfig. This required extending said function to take the
prepare_custom_config_dict as an optional argument.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

HDCharles committed:
[AO][bugfix] Fixing FX QAT but for untraceable modules

Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by setting qconfig=None for children of
untraceable modules. New functionality was needed in QuantizationTracer
to distinguish these modules from custom and standalone modules which
do not need such treatment.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles committed:
Update on "[AO][bugfix] Fixing FX QAT but for untraceable modules"


Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by setting qconfig=None for children of
untraceable modules. New functionality was needed in QuantizationTracer
to distinguish these modules from custom and standalone modules which
do not need such treatment.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

jerryzh168 commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -134,13 +139,22 @@ def __init__(
+        ):
+            for name, child in m.named_children():
+                if not hasattr(child, "qconfig"):
+                    child.qconfig = None  # type: ignore[assignment]
```
I feel ideally we should set qconfig in one place, to keep code easier to maintain and reason about, can we do this in prepare?

HDCharles committed:
Update on "[AO][bugfix] Fixing FX QAT but for untraceable modules"


Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by setting qconfig=None for children of
untraceable modules. New functionality was needed in QuantizationTracer
to distinguish these modules from custom and standalone modules which
do not need such treatment.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

HDCharles committed:
Update on "[AO][bugfix] Fixing FX QAT but for untraceable modules"


Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by skipping the children of untraceable modules during
propagate_qconfig. This required extending said function to take the
prepare_custom_config_dict as an optional argument.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

vkuzo commented on torch/ao/quantization/quantize.py:
```diff
@@ -61,10 +64,15 @@ def _propagate_qconfig_helper(module, qconfig_dict,
         module_prefix = prefix + '.' + name if prefix else name
-        _propagate_qconfig_helper(child, qconfig_dict,
-                                  qconfig_with_device_check, module_prefix)
+        if prepare_custom_config_dict is None or not (
```
can we add a comment with a human readable description on what this is doing?

vkuzo commented on torch/ao/quantization/quantize.py:
```diff
@@ -61,10 +64,15 @@ def _propagate_qconfig_helper(module, qconfig_dict,
+            )
 
-def propagate_qconfig_(module, qconfig_dict=None):
+def propagate_qconfig_(module, qconfig_dict=None, prepare_custom_config_dict=None):
```
could we add the arg to the docstring too?

vkuzo reviewed (approved):
looks great!

jerryzh168 commented:
looks great, do we want to add support for custom module as well in a followup PR?

HDCharles committed:
Update on "[AO][bugfix] Fixing FX QAT but for untraceable modules"


Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by skipping the children of untraceable modules during
propagate_qconfig. This required extending said function to take the
prepare_custom_config_dict as an optional argument.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

HDCharles committed:
Update on "[AO][bugfix] Fixing FX QAT but for untraceable modules"


Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by skipping the children of untraceable modules during
propagate_qconfig. This required extending said function to take the
prepare_custom_config_dict as an optional argument.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).

HDCharles committed:
Update on "[AO][bugfix] Fixing FX QAT but for untraceable modules"


Summary: see issue: https://github.com/pytorch/pytorch/issues/74240

this fixes that issue by skipping the children of untraceable modules during
propagate_qconfig. This required extending said function to take the
prepare_custom_config_dict as an optional argument.

Test Plan:python test/test_quantization.py
python test/test_quantization.py TestQuantizeFx.test_qat_skip_untraced

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34916074](https://our.internmc.facebook.com/intern/diff/D34916074)

[ghstack-poisoned]

HDCharles commented:
@HDCharles has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34916074).
